rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,840 @@
+            let mut skipped = false;
+            SCLogDebug!("response: handling {}", current.len());
+            match parse_message(current, self.protocol_version, self.max_msg_len as usize) {
+                Ok((rem, msg)) => {
+                    SCLogDebug!("response msg {:?}", msg);
+                    if let MQTTOperation::TRUNCATED(ref trunc) = msg.op {
+                        SCLogDebug!("found truncated with skipped {} current len {}", trunc.skipped_length, current.len());
+                        if trunc.skipped_length >= current.len() {
+                            skipped = true;
+                            self.skip_response = trunc.skipped_length - current.len();
+                        } else {
+                            current = &current[trunc.skipped_length..];
+                            self.skip_response = 0;
+                        }
+                        SCLogDebug!("skip_response now {}", self.skip_response);
+                    }
+                    self.handle_msg(msg, true);
+                    if skipped {
+                        return AppLayerResult::ok();
+                    }
+                    consumed += current.len() - rem.len();
+                    current = rem;
+                }
+                Err(nom::Err::Incomplete(_)) => {
+                    SCLogDebug!("incomplete response: consumed {} needed {} (input len {})", consumed, (current.len() + 1), input.len());
+                    return AppLayerResult::incomplete(consumed as u32, (current.len() + 1) as u32);
+                }
+                Err(_) => {
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        return AppLayerResult::ok();
+    }
+
+    fn set_event(tx: &mut MQTTTransaction, event: MQTTEvent) {
+        let ev = event as u8;
+        core::sc_app_layer_decoder_events_set_event_raw(&mut tx.events, ev);
+    }
+
+    fn tx_iterator(
+        &mut self,
+        min_tx_id: u64,
+        state: &mut u64,
+    ) -> Option<(&MQTTTransaction, u64, bool)> {
+        let mut index = *state as usize;
+        let len = self.transactions.len();
+
+        while index < len {
+            let tx = &self.transactions[index];
+            if tx.tx_id < min_tx_id + 1 {
+                index += 1;
+                continue;
+            }
+            *state = index as u64;
+            return Some((tx, tx.tx_id - 1, (len - index) > 1));
+        }
+
+        return None;
+    }
+}
+
+// C exports.
+
+export_tx_get_detect_state!(rs_mqtt_tx_get_detect_state, MQTTTransaction);
+export_tx_set_detect_state!(rs_mqtt_tx_set_detect_state, MQTTTransaction);
+
+#[no_mangle]
+pub extern "C" fn rs_mqtt_probing_parser(
+    _flow: *const Flow,
+    _direction: u8,
+    input: *const u8,
+    input_len: u32,
+    _rdir: *mut u8,
+) -> AppProto {
+    let buf = build_slice!(input, input_len as usize);
+    match parse_fixed_header(buf) {
+        Ok(_) => unsafe { ALPROTO_MQTT },
```

Comment:
I mean, will not `parse_fixed_header` always return `Ok` if it has enough data ?

Reply:
Well, if we try to parse the full message, we might run into a situation where we don't have enough data to parse a full message, what then? I was under the impression that the probing code only gives a fraction of the traffic to the probing parser, to verify that it could indeed be the protocol specified by the defined port. How much do we get? What if the first message is too long to parse, do we need to replicate the skipping code in the probing parser as well?
I think it should be enough to:

- make sure that there is enough data, and
- make sure that the message type and QOS level makes sense in the context of the protocol (I will add extra checks for that).

Given that the port is already quite specific for MQTT, I don't really see any pressing need to be be more specific than that during probing. Otherwise we would need more complexity here and also might be parsing things redundantly multiple times.

Reply:
I do not mean to parse the full message.
This probing code should just look at the header I think.

> make sure that the message type and QOS level makes sense in the context of the protocol (I will add extra checks for that).

Yes, that is what I was looking for.
I will be happy when those checks are added

Reply:
Okay, will do :+1: